# 1. Hafta: Google Trends Veri Toplama ve İnceleme (pytrends)

Bu notebook'ta `pytrends` kütüphanesini kullanarak Google Trends üzerinden belirlenen teknolojilerin tarihsel arama hacimlerini (Time Series) çekecek ve inceleyeceğiz.

### Takip Edilecek Teknolojiler:
* **AI Agents** (Yapay Zeka Ajanları)
* **LangChain** (LLM Uygulama Geliştirme Framework'ü)
* **LLM** (Büyük Dil Modelleri)

In [5]:
from pytrends.request import TrendReq
# TrendReq, pytrends kütüphanesinin Google Trends'e istek (request) gönderen sınıfıdır (class).
#Yani Google Trends ile iletişimi başlatır.

dir(TrendReq)


['CATEGORIES_URL',
 'ERROR_CODES',
 'GENERAL_URL',
 'GET_METHOD',
 'GetGoogleCookie',
 'GetNewProxy',
 'INTEREST_BY_REGION_URL',
 'INTEREST_OVER_TIME_URL',
 'MULTIRANGE_INTEREST_OVER_TIME_URL',
 'POST_METHOD',
 'REALTIME_TRENDING_SEARCHES_URL',
 'RELATED_QUERIES_URL',
 'SUGGESTIONS_URL',
 'TODAY_SEARCHES_URL',
 'TOP_CHARTS_URL',
 'TRENDING_SEARCHES_URL',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__firstlineno__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__static_attributes__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_get_data',
 '_tokens',
 'build_payload',
 'categories',
 'get_historical_interest',
 'interest_by_region',
 'interest_over_time',
 'multirange_interest_over_time',
 'realtime_trending_searches',
 'related_qu

In [7]:
help(TrendReq.build_payload)

Help on function build_payload in module pytrends.request:

build_payload(self, kw_list, cat=0, timeframe='today 5-y', geo='', gprop='')
    Create the payload for related queries, interest over time and interest by region



In [14]:
pytrends = TrendReq()

# 1. Google Trends bağlantısını başlatıyoruz
# hl='en-US' -> Arama sonuçlarının dilini İngilizce ayarlar
# tz=360     -> Saat dilimini (Timezone) ayarlar
pytrend = TrendReq(hl='en-US', tz=180)

# 2. İncelemek istediğimiz kelimeleri liste olarak tanımlıyoruz
x= ["Java", "Python", "C++"]

# 3. Parametreleri(istek) Google Trends'e gönderiyoruz (Payload oluşturma)
pytrend.build_payload(kw_list= x, timeframe="today 5-y", geo= "TR")

In [ ]:
help(pytrend.interest_over_time) # o isteiğimiz verileri getir 

Help on method interest_over_time in module pytrends.request:

interest_over_time() method of pytrends.request.TrendReq instance
    Request data from Google's Interest Over Time section and return a dataframe



# method of TrendReq instance

Bu cümle çok önemli.

Diyor ki:

Bu, TrendReq sınıfından oluşturulmuş bir nesnenin metodudur.

# "Interest" ne demek?

Burada küçük ama önemli bir yanlış anlaşılma oluyor.

Google sana şunu söylemiyor:

"Python kelimesi 1.235.678 kez arandı."

Hayır.

Google Trends arama sayısını vermez.

Onun yerine 0 ile 100 arasında normalize edilmiş bir ilgi puanı verir.

In [15]:
df = pytrend.interest_over_time()

print(df)

            Java  Python  C++  isPartial
date                                    
2021-08-01     8       7   58      False
2021-08-08     8       7   58      False
2021-08-15     9       8   61      False
2021-08-22     9       8   65      False
2021-08-29     9       7   57      False
...          ...     ...  ...        ...
2026-07-05    12      13   71      False
2026-07-12    12      12   67      False
2026-07-19    10       8   66      False
2026-07-26    11      10   68      False
2026-08-02    11      15   72       True

[262 rows x 4 columns]


### TrendReq timezone notu

Türkiye için:

```python
pytrend = TrendReq(
    hl="tr-TR",
    tz=180
)

In [16]:
df.head()

,Java,Python,C++,isPartial
date,,,,
2021-08-01,8,7,58,False
2021-08-08,8,7,58,False
2021-08-15,9,8,61,False
2021-08-22,9,8,65,False
2021-08-29,9,7,57,False


- date burda index
- Bu:"C++ 65 kez, Java 9 kez arandı." demek değil.
- Bu sorgunun Google Trends ölçeğine göre C++'ın o hafta çok daha yüksek göreli ilgiye sahip olduğunu gösteriyor.

# isPartial nedir?

Bu da önemli.

- False

şunu gösterir:

Bu zaman diliminin verisi tamamlanmış.

Örneğin içinde bulunduğumuz hafta henüz bitmediyse Google son satırda:

- True

verebilir.

Çünkü o haftanın verisi henüz tamamlanmamıştır.

Bu özellikle model eğitirken önemli. Çünkü tamamlanmamış son haftayı kullanmak yanıltıcı olabilir.



In [17]:
df. shape

(262, 4)

In [18]:
df.info

<bound method DataFrame.info of             Java  Python  C++  isPartial
date                                    
2021-08-01     8       7   58      False
2021-08-08     8       7   58      False
2021-08-15     9       8   61      False
2021-08-22     9       8   65      False
2021-08-29     9       7   57      False
...          ...     ...  ...        ...
2026-07-05    12      13   71      False
2026-07-12    12      12   67      False
2026-07-19    10       8   66      False
2026-07-26    11      10   68      False
2026-08-02    11      15   72       True

[262 rows x 4 columns]>

In [19]:
df.tail()

,Java,Python,C++,isPartial
date,,,,
2026-07-05,12,13,71,False
2026-07-12,12,12,67,False
2026-07-19,10,8,66,False
2026-07-26,11,10,68,False
2026-08-02,11,15,72,True


In [20]:
df.describe()

,Java,Python,C++
count,262.000000,262.000000,262.000000
mean,11.049618,12.236641,72.877863
std,1.573676,2.792524,7.964503
min,7.000000,7.000000,54.000000
25%,10.000000,10.250000,67.000000
50%,11.000000,12.000000,72.000000
75%,12.000000,14.000000,79.000000
max,16.000000,22.000000,100.000000


Bunu şöyle okuyacaksın:

count = 262

- Her sütunda 262 değer var. Yani burada ilk bakışta eksik değer görünmüyor.

mean ortalama demek.

Senin verinde:

Java   ≈ 11.05
Python ≈ 12.24
C++    ≈ 72.88

- Yani 5 yıllık dönem genelinde C++ açık ara daha yüksek çıkmış.

Sen başta Python demiştin ama veri bize:

C++ > Python > Java

diyor.

In [21]:
df.info()

<class 'pandas.DataFrame'>
DatetimeIndex: 262 entries, 2021-08-01 to 2026-08-02
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   Java       262 non-null    int64
 1   Python     262 non-null    int64
 2   C++        262 non-null    int64
 3   isPartial  262 non-null    bool 
dtypes: bool(1), int64(3)
memory usage: 8.4 KB


Önce şu hatayı netleştireyim:

df.info

yazarsan Python sana fonksiyonun kendisini gösterir:

<bound method DataFrame.info of ...>

Çünkü parantez koymadın. Yani “info metodunu çalıştır” demedin, sadece “info metodu nedir?” dedin.

Doğrusu:

df.info()

Bunu çalıştırınca şöyle bir şey görmen gerekir:

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 262 entries, ...
Data columns (total 4 columns):
 ...

Yani küçük ama çok önemli kural:

df.info      # fonksiyonun kendisi
df.info()    # fonksiyonu çalıştır

In [23]:
df.index

df.dtypes

Java         int64
Python       int64
C++          int64
isPartial     bool
dtype: object

In [24]:
df.isnull().sum()

Java         0
Python       0
C++          0
isPartial    0
dtype: int64

In [25]:
df.describe.__module__

'pandas.core.generic'

In [26]:
help(df.index)

Help on DatetimeIndex in module pandas object:

class DatetimeIndex(pandas.core.indexes.datetimelike.DatetimeTimedeltaMixin)
 |  DatetimeIndex(
 |      data=None,
 |      freq: Frequency | lib.NoDefault = <no_default>,
 |      tz=<no_default>,
 |      ambiguous: TimeAmbiguous = 'raise',
 |      dayfirst: bool = False,
 |      yearfirst: bool = False,
 |      dtype: Dtype | None = None,
 |      copy: bool | None = None,
 |      name: Hashable | None = None
 |  ) -> Self
 |
 |  Immutable ndarray-like of datetime64 data.
 |
 |  Represented internally as int64, and which can be boxed to Timestamp objects
 |  that are subclasses of datetime and carry metadata.
 |
 |  .. versionchanged:: 2.0.0
 |      The various numeric date/time attributes (:attr:`~DatetimeIndex.day`,
 |      :attr:`~DatetimeIndex.month`, :attr:`~DatetimeIndex.year` etc.) now have dtype
 |      ``int32``. Previously they had dtype ``int64``.
 |
 |  Parameters
 |  ----------
 |  data : array-like (1-dimensional)
 |      Dat

bu son iki kod herhangi bir metodun hangi paket, hangi sınıf falan gösteriyo

In [27]:
df["isPartial"].value_counts()
# Bu sana kaç tane: False True olduğunu gösterecek.Böylece incomplete/kısmi veri kaç tane, öğreniyoruz.

isPartial
False    261
True       1
Name: count, dtype: int64

In [28]:
df.dtypes

Java         int64
Python       int64
C++          int64
isPartial     bool
dtype: object

Anlamları:

Java        → integer
Python      → integer
C++         → integer
isPartial   → boolean

dtype: object ise "bütün DataFrame object" demek değil. Burada gördüğün sonuç (df.dtypes) kendisi bir Pandas Series ve onun dtype'ı object. Bu ufak ayrıntı ileride kafa karıştırmasın.

In [29]:
df.index.to_series().diff()

date
2021-08-01      NaT
2021-08-08   7 days
2021-08-15   7 days
2021-08-22   7 days
2021-08-29   7 days
              ...  
2026-07-05   7 days
2026-07-12   7 days
2026-07-19   7 days
2026-07-26   7 days
2026-08-02   7 days
Name: date, Length: 262, dtype: timedelta64[s]

# .diff()

Bir değer ile kendisinden önceki değer arasındaki farkı hesaplar.

Mesela:

1 Ağustos
8 Ağustos
15 Ağustos
22 Ağustos

için sonuç:

NaT
7 days
7 days
7 days

İlk satırda NaT olması normal; kendisinden önce karşılaştırabileceği tarih yok.

NaT kabaca datetime dünyasının NaN'ı gibi düşünebilirsin: Not a Time.

Bizim beklediğimiz frekans haftalık olduğu için normal fark:

7 days

olmalı.

Eğer bir yerde:

14 days

görürsek:

🚨 Arada muhtemelen bir haftalık timestamp eksik.

In [ ]:
df.index.to_series().diff().value_counts()
#Elimizde 262 tarih var. Ardışık tarihler arasında ise 261 fark bulunuyor ve bu 261 farkın tamamı 7 gün. yani eksik tarih yok. sadece 1 tane son hafta tamamlanmamış.

date
7 days    261
Name: count, dtype: int64

# Df' i kopyalayıp eksiltelim ve linner interpolationı test edelim
gerçek pytrends verisindeki gerçek bir değeri bilerek sileceğiz → interpolation ile tahmin edeceğiz → gerçek değerle tahmini karşılaştıracağız

In [31]:
test_df= df.copy()

In [32]:
help(df.copy)

Help on method copy in module pandas.core.generic:

copy(deep: bool = True) -> Self method of pandas.DataFrame instance
    Make a copy of this object's indices and data.

    When ``deep=True`` (default), a new object will be created with a
    copy of the calling object's data and indices. Modifications to
    the data or indices of the copy will not be reflected in the
    original object (see notes below).

    When ``deep=False``, a new object will be created without copying
    the calling object's data or index (only references to the data
    and index are copied). With Copy-on-Write, changes to the original
    will *not* be reflected in the shallow copy (and vice versa). The
    shallow copy uses a lazy (deferred) copy mechanism that copies the
    data only when any changes to the original or shallow copy are made,
    ensuring memory efficiency while maintaining data integrity.

    .. note::
        In pandas versions prior to 3.0, the default behavior without
        Copy

In [51]:
import numpy as np

# 1) Orijinal veriyi bozmamak için kopyasını oluştur
test_df = df.copy()

# 2) Test edeceğimiz gerçek değeri sakla
gerçek_deger = test_df.loc["2024-05-05", "Python"]

# 3) O değeri bilerek eksik hale getir
test_df.loc["2024-05-05", "Python"] = np.nan

# 4) Eksik halini göster
print("=== INTERPOLATION ÖNCESİ ===")
print(test_df.loc["2024-04-28":"2024-05-12", "Python"])

# 5) Python sütunundaki eksik değerleri linear interpolation ile doldur
test_df["Python"] = test_df["Python"].interpolate(method="linear")

# 6) Doldurulmuş halini göster
print("\n=== INTERPOLATION SONRASI ===")
print(test_df.loc["2024-04-28":"2024-05-12", "Python"])

# 7) Interpolation'ın tahmin ettiği değeri al
tahmin = test_df.loc["2024-05-05", "Python"]

# 8) Gerçek değer ile tahmini karşılaştır
print("\n=== SONUÇ ===")
print("Gerçek değer:", gerçek_deger)
print("Tahmin edilen:", tahmin)
print("Hata:", abs(gerçek_deger - tahmin))

=== INTERPOLATION ÖNCESİ ===
date
2024-04-28    13.0
2024-05-05     NaN
2024-05-12    14.0
Name: Python, dtype: float64

=== INTERPOLATION SONRASI ===
date
2024-04-28    13.0
2024-05-05    13.5
2024-05-12    14.0
Name: Python, dtype: float64

=== SONUÇ ===
Gerçek değer: 14
Tahmin edilen: 13.5
Hata: 0.5


# loc
df.loc[satır, sütun]

kullanıyoruz.

Örneğin:

df.loc["2024-05-05", "Python"]

şöyle okunur:

df DataFrame'inde
"2024-05-05" satırına git,
"Python" sütunundaki değeri getir.

# Peki neden NumPy'ye ihtiyacımız var?

Biz bir hücreyi:

"Burada veri yok."

durumuna getirmek istiyoruz.

Bunun için:

np.nan

kullanabiliriz.

NaN ne?

Not a Number

Veri analizinde kabaca:

Eksik sayısal değer

anlamında kullanılır.

1. # interpolate() nedir?

interpolate() bir Pandas metodudur.

Yani şu yapıdan gelir:

Pandas
  ↓
Series / DataFrame
  ↓
interpolate()

Görevi:

Eksik değerleri, çevresindeki bilinen değerlere bakarak tahmin edip doldurmak.

Mesela:

10
NaN
20

için lineer interpolation:

10
15
20

yapabilir.

In [35]:
test_df["Python"].interpolate()

date
2021-08-01     7
2021-08-08     7
2021-08-15     8
2021-08-22     8
2021-08-29     7
              ..
2026-07-05    13
2026-07-12    12
2026-07-19     8
2026-07-26    10
2026-08-02    15
Name: Python, Length: 262, dtype: int64

# Bu ne yapıyor?

test_df içinden yalnızca:

Python

sütununu seçiyor.

Artık elimizde bütün tablo değil, tek sütunluk bir Pandas Series var.

Örneğin:

date
2024-04-28    10
2024-05-05    NaN
2024-05-12    14

- Burada henüz test_df'yi değiştirmiyoruz.

Pandas sadece:

"Doldursaydım sonuç böyle olurdu."

diye bize yeni sonucu gösteriyor.

Bunu çalıştırıp özellikle sildiğin tarihe bak.

In [1]:
df.head()

NameError: name 'df' is not defined